### ISSUE TITLE
awesomebar: Store history and session in indexeddb

### ISSUE THREAD

**Issue by paulrouget** [mozilla/browser.html#11]
We need to be able to store all the history in the DB (datastore/indexeddb), and also have a mechanism to query the db, including suggestions from search engines.

**Comment by paulrouget** [comment]
_From @giuscri on December 11, 2014 9:51_

What about using `localForage` APIs instead? Code must be simpler to write that way.
```
// Use IndexedDB as the backend of localForage
localforage.setDriver(localforage.IndexedDB);

// ...

// Setting 'google.com' once visited
localforage.getItem('google.com', function (err, value) {
    if (!value) { localforage.setItem('google.com', true, null); }
});
```
Using Promises should be a possibility too.

**Comment by paulrouget** [comment]
I'm not against it. What would be the strategy to query entries? For example, if the user types:

> France

We need to be able to query all the entries with a URLs and titles that include "France".

**Comment by paulrouget** [comment]
_From @giuscri on December 11, 2014 11:16_

Sounds pretty dirty, but `SELECT`ing something needs to loop over all the tuples in any case, so ... using `localforage.iterate`?
```
var url = ...;
localforage.iterate(function (value, host) {
    // If url's host and host are
    // the same, update awesomebar
    //...
}, null);
```
Even if probably `(host,value)` is not the base design to think about; probably the naive `(id,host)` is better.

**Comment by paulrouget** [comment]
_From @Yoric on December 11, 2014 12:21_

Unfortunately, `Promise` don't work nicely with indexedDB. You need to be extremely careful when using a `Promise` if you want to be sure that you are not inadvertently leaving a transaction mid-way.

Also, I'm not sure that localforage would be very useful. The main advantage of this library, iirc, is that it can fallback to localStorage or even cookies if indexedDB is not present. This is not a setting in which we're planning to run firefox.html.

Consequently, I would go for raw indexedDB, or perhaps asyncStorage (which I haven't evaluated myself) if we wish a simpler API.

**Comment by paulrouget** [comment]
@Yoric Can you take a look at asyncStorage and tell us if it's a viable option? This bug is, I think, the last piece to make htmlrunner a usable browser.

**Comment by paulrouget** [comment]
I think we should just use `places.js` from B2G.

**Comment by paulrouget** [comment]
This is how it works in Gaia:

- save: https://github.com/mozilla-b2g/gaia/blob/master/apps/system/js/places.js
- query: https://github.com/mozilla-b2g/gaia/blob/master/apps/search/js/places_idb.js

It requires datastore. To enable datastore, the runtime needs to enable datastores: `+pref("dom.datastore.enabled", true);` and firefox.html needs to ask for the datastore permission in the manifest:  
```
 "datastores-access": {
   "places": {
     "access": "readonly",
     "description": "Stores data about browsing history."
   },
   "bookmarks_store": {
     "access": "readonly",
     "description": "Stores data about bookmarks"
   }
```

**Comment by Gozala** [comment]
First of all I would really like to use "append only" strategy for data store, the wisdom of this strategy is explained very well in by ["Accountants don't use erasers"](http://blogs.msdn.com/b/pathelland/archive/2007/06/14/accountants-don-t-use-erasers.aspx). Given the nature of browser history "append only" strategy is quit a good fit and could let us avoid any sync conflicts.

With that in mind I have evaluated several options, proposed here and in other communication channels. I'd like to list pros and cons of each approach below as I see them

## [PouchDB](http://pouchdb.com/)

### Pros

1. Use standard indexed db API.
1. Can sync with couch db so we can get sync for free.
1. Simpler API then raw indexed db.
1. Can easily be used with "append only" strategy, & local (non synced) document can be created for caching.

### Cons

1. As far as I can tell there is no way to select records by index / get a cursor by index, which potentially will make queries slow as DB will grow in size.
1. We'll are also kind of buying into couchdb.

### [Raw Indexed DB](https://developer.mozilla.org/en-US/docs/Web/API/IndexedDB_API/Using_IndexedDB)


### Pros

1. Standard HTML API.
1. We could easily use "append only" strategy and store aggregated visit counts in second db.
1. Searches can be very fast as we can get cursors for indexes.

### Cons

1.  We would have to implement sync ourselves.


### [Gaia Places API](https://github.com/mozilla-b2g/gaia/blob/master/apps/system/js/places.js)

### Pros

1. Can be easily integrated with our code base.

### Cons

1. They use non standard (at least not yet) and less mature [Data Store](https://developer.mozilla.org/en-US/docs/Web/API/Data_Store_API)
2. It also depends on some `asyncStorage` that I'm not sure I know where it's coming form.
3. My understanding of Data Store API is it's primarily designed to allow sharing across the apps, which I don't think is very relevant in our case.
4. Data Store API does not seems to have any indexing feature.

**Comment by Gozala** [comment]
At this point I plan to try a PouchDB as it's well tested & maintained piece of software that also solves bring your data with you issue. We could always reconsider that and switch to raw indexedb or even Data Store if we find other options more appropriate.

**Comment by donaldpipowitch** [comment]
Looking forward to some insight into PouchDB. I heard nothing but good about PouchDB, but haven't used it myself yet. It sounds like a good candidate for the history.

**Comment by paulrouget** [comment]
`asyncStorage`: https://github.com/mozilla-b2g/gaia/blob/master/shared/js/async_storage.js

> My understanding of Data Store API is it's primarily designed to allow sharing across the apps, which I don't think is very relevant in our case.

Sharing data across **processes**. IndexedDB doesn't allow that. Keep that in mind. But if there anything we want to share across processes, we can always use the settings API, or use a datastore fed from indexeddb.

> At this point I plan to try a PouchDB as it's well tested & maintained piece of software that also solves bring your data with you issue.

Agreed.

If you can implement all that into a shared workers, that would be pretty awesome.

**Comment by daleharvey** [comment]
@Gozala pinged me on this

I wrote the awesome bar for fxos and maintain pouchdb, so I know a little about both ends, will explain how it works with fxos. We use DataStore specifically to share the data across applications, the system app is the one who gets the visited history, but the search app is the one that displays it, so we record visits in the system app (https://github.com/mozilla-b2g/gaia/blob/master/apps/system/js/places.js) in around the format:

```
{
  url: 'http://....',
  frecency: X, // Right now this is purely frequency, we dont decay, but will likely do
  title: 'title'
  visits: [], // Array of timestamps
  icons: {} // Metadata about the icons 
}
```

Then in the search app, we read the 'stream' from datastore and store that locally into 2 seperate idb stores, one is visits, indexed by the time, the other is urls, indexed by frecency, when we display history we read from visits, when we show the awesome bar we run a cursor through the frecency index until we find enough matching urls (done in https://github.com/mozilla-b2g/gaia/blob/master/apps/search/js/places_idb.js)

**Comment by gordonbrander** [comment]
Important for dogfooding, but not really for demos.

**Comment by Gozala** [comment]
Fixed by #373

**Comment by gordonbrander** [comment]
:+1:
